electron_app/DownloadManager.js

```diff
@@ -1,25 +1,48 @@
 const Downloader = require('./Downloader');
 const path = require('path');
-const { handleFolder } = require('./helpers/handleDisk');
+const { handleFolder, fileAlreadyExists } = require('./helpers/handleDisk');
 const downloadStats = require('./helpers/downloadInfo').downloadStats;
 const handleEvents = require('./helpers/handleEvents');
 
 let filePath = '';
 let allDownloads = [];
+let fileExists = true;
 
-const downloadManager = (items, access_token, win) => {
+const downloadManager = (items, access_token, electronWin) => {
   items.forEach(item => {
-    if (item.preserveStructure) {
-      const structurePath = item.path.substring(item.path.lastIndexOf('/'), 0);
-      handleFolder(path.join(item.destination, structurePath)).then(result => {
-        allDownloads.push(processDownload(access_token, item, result, win));
-      });
-    } else {
-      allDownloads.push(processDownload(access_token, item, item.destination, win));
-    }
+    let count = 0;
+    let fileName = item.name;
+    do {
+      fileExistsInDestination(item, fileName);
+      if (fileExists) {
+        count++;
+        fileName = item.name.substring(0, item.name.lastIndexOf('.')) + '(' + (count) + ')' +
+          item.name.substring(item.name.lastIndexOf('.'));
+      } else {
+        item.name = fileName;
+        processItem(item, access_token, electronWin);
+      }
+    } while (fileExists);
   });
 };
 
+const fileExistsInDestination = (item, nameCheck) => {
+  fileExists = item.preserveStructure ?
+    fileAlreadyExists(path.join(item.destination + '/' + item.path.substring(item.path.lastIndexOf('/'), 0) + '/' + nameCheck))
+    : fileExists = fileAlreadyExists(path.join(item.destination + '/' + nameCheck));
+};
+
+const processItem = (item, token, win) => {
+  if (item.preserveStructure) {
+    const structurePath = item.path.substring(item.path.lastIndexOf('/'), 0);
```

Comment:
It would be nice if there was a way to avoid testing for `preserveStructure` twice. It also duplicates the way the path is constructed. Since the relative path doesn't change on each loop, only the filename, the destination directory can be computed once before the `do` loop on line 15 above. That would avoid the need to check `preserveStructure` in the loop.

Reply:
Thanks for the advice, I removed that verification from the loop, and item's destination path is only constructed once, changing each file's name if necessary.